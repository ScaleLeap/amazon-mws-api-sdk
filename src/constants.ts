/**
 * The version value is automatically replaced during builds.
 */
export const USER_AGENT = '@scaleleap/amazon-mws-api-sdk/1.9.0'
