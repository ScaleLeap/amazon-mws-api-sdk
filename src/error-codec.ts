import { Codec, exactly, oneOf, optional, string } from 'purify-ts/Codec'

export enum CodeEnum {
  InputStreamDisconnected,
  InvalidParameterValue,
  AccessDenied,
  InvalidAccessKeyId,
  SignatureDoesNotMatch,
  InvalidAddress,
  InternalError,
  'Internal Error',
  QuotaExceeded,
  RequestThrottled,
  ResourceNotFound,
  ScheduledPackageAlreadyExists,
  RegionNotSupported,
  ScheduleWindowExpired,
  InvalidOrderState,
  PickupSlotNotAvailable,
  AccessToFeedProcessingResultDenied,
  ContentMD5Missing,
  ContentMD5DoesNotMatch,
  FeedCanceled,
  FeedProcessingResultNoLongerAvailable,
  FeedProcessingResultNotReady,
  InputDataError,
  InvalidFeedSubmissionId,
  InvalidFeedType,
  InvalidRequest,
  NonRetriableInternalError,
  RetriableInternalError,
  AccessToReportDenied,
  InvalidReportId,
  InvalidReportRequestId,
  InvalidReportType,
  InvalidScheduleFrequency,
  ReportNoLongerAvailable,
  ReportNotReady,
  DependencyFatalException,
  DependencyRetriableException,
  DependencyUnauthorizedException,
  InternalErrorFatalException,
  InvalidInputFatalException,
}

export const Error = Codec.interface({
  Type: string,
  Code: oneOf(Object.keys(CodeEnum).map((x) => exactly(x)) as [Codec<string>, ...Codec<string>[]]),
  Message: string,
  Detail: optional(string),
})

export const MWSApiError = Codec.interface({
  ErrorResponse: Codec.interface({
    Error,
    RequestID: optional(string),
    RequestId: optional(string),
  }),
})
