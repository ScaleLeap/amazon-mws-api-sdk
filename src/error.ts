/* eslint-disable max-classes-per-file */
export class MWSError extends Error {
  constructor(...parameters: string[]) {
    // Propagate some vendor-specific arguments
    super(...parameters)

    // If someone downlevels the compilation target to es5
    Object.setPrototypeOf(this, MWSError.prototype)

    // Maintains proper stack trace (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, MWSError)
    }
  }
}

export class HttpError extends MWSError {
  public message = 'Encountered an error while sending a request'

  constructor(public error: unknown, ...parameters: string[]) {
    super(...parameters)
    Object.setPrototypeOf(this, HttpError.prototype)

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, HttpError)
    }
  }
}

export class ParsingError extends MWSError {
  public message = 'Encountered an error while parsing a response'

  constructor(public error: string, ...parameters: string[]) {
    super(...parameters)
    Object.setPrototypeOf(this, ParsingError.prototype)

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ParsingError)
    }
  }
}
/* eslint-enable max-classes-per-file */
