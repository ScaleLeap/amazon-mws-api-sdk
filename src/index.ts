export * from './http'
export * from './mws'
export * from './sections/sellers'
export * from './sections/orders'
export * from './sections/products/products'
export * from './sections/fulfillment-inventory'
export * from './sections/reports'
export * from './error'
export * from '@scaleleap/amazon-marketplaces'
