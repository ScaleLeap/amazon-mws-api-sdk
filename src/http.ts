import { AmazonMarketplace } from '@scaleleap/amazon-marketplaces'
import axios from 'axios'
import parser from 'fast-xml-parser'
import { URLSearchParams } from 'url'

import {
  AccessDenied,
  AccessToFeedProcessingResultDenied,
  AccessToReportDenied,
  ContentMD5DoesNotMatch,
  ContentMD5Missing,
  DependencyFatalException,
  DependencyRetriableException,
  DependencyUnauthorizedException,
  enhanceError,
  FeedCanceled,
  FeedProcessingResultNoLongerAvailable,
  FeedProcessingResultNotReady,
  InputDataError,
  InputStreamDisconnected,
  InternalError,
  InternalErrorFatalException,
  InvalidAccessKeyId,
  InvalidAddress,
  InvalidFeedSubmissionId,
  InvalidFeedType,
  InvalidInputFatalException,
  InvalidOrderState,
  InvalidParameterValue,
  InvalidReportId,
  InvalidReportType,
  InvalidRequest,
  InvalidScheduleFrequency,
  InvalidUPCIdentifier,
  NonRetriableInternalError,
  PickupSlotNotAvailable,
  QuotaExceeded,
  RegionNotSupported,
  ReportNoLongerAvailable,
  ReportNotReady,
  RequestThrottled,
  ResourceNotFound,
  RetriableInternalError,
  ScheduledPackageAlreadyExists,
  ScheduleWindowExpired,
  SignatureDoesNotMatch,
} from './error'
import { MWSApiError } from './error-codec'
import { sign } from './sign'

export interface MWSOptions {
  marketplace: AmazonMarketplace
  awsAccessKeyId: string
  mwsAuthToken: string
  sellerId: string
  secretKey: string
}

type HttpMethod = 'GET' | 'POST'
type ParameterTypes = string | number | (number | string)[] | object[] | boolean | undefined
type Parameters = Record<string, ParameterTypes>
type CleanParameters = Record<string, string>

export enum Resource {
  Sellers = 'Sellers',
  Orders = 'Orders',
  Products = 'Products',
  FulfilmentInventory = 'FulfillmentInventory',
  Report = 'Report',
}

interface ResourceActions {
  [Resource.Sellers]:
    | 'ListMarketplaceParticipations'
    | 'ListMarketplaceParticipationsByNextToken'
    | 'GetServiceStatus'
  [Resource.Orders]:
    | 'ListOrders'
    | 'ListOrdersByNextToken'
    | 'GetOrder'
    | 'ListOrderItems'
    | 'ListOrderItemsByNextToken'
    | 'GetServiceStatus'
  [Resource.Products]:
    | 'ListMatchingProducts'
    | 'GetMatchingProduct'
    | 'GetMatchingProductForId'
    | 'GetCompetitivePricingForSKU'
    | 'GetCompetitivePricingForASIN'
    | 'GetLowestOfferListingsForSKU'
    | 'GetLowestOfferListingsForASIN'
    | 'GetLowestPricedOffersForSKU'
    | 'GetLowestPricedOffersForASIN'
    | 'GetMyFeesEstimate'
    | 'GetMyPriceForSKU'
    | 'GetMyPriceForASIN'
    | 'GetProductCategoriesForSKU'
    | 'GetProductCategoriesForASIN'
    | 'GetServiceStatus'
  [Resource.FulfilmentInventory]:
    | 'ListInventorySupply'
    | 'ListInventorySupplyByNextToken'
    | 'GetServiceStatus'
  [Resource.Report]:
    | 'RequestReport'
    | 'GetReportRequestList'
    | 'GetReportRequestListByNextToken'
    | 'GetReportRequestCount'
    | 'CancelReportRequests'
    | 'GetReportList'
    | 'GetReportListByNextToken'
    | 'GetReportCount'
    | 'GetReport'
    | 'ManageReportSchedule'
    | 'GetReportScheduleList'
    | 'GetReportScheduleListByNextToken'
    | 'GetReportScheduleCount'
    | 'UpdateReportAcknowledgements'
}

interface Request {
  url: string
  method: HttpMethod
  headers: Record<string, string>
  data?: string
}

interface ResourceInfo<TResource extends Resource> {
  resource: TResource
  version: string
  action: ResourceActions[TResource]
  parameters: Parameters
}

export interface RequestMeta {
  requestId: string
  timestamp: Date
  quotaMax: number
  quotaRemaining: number
  quotaResetOn: Date
}

interface RequestResponse {
  data: string
  headers: Record<string, string>
}

const canonicalizeParameters = (parameters: CleanParameters): string => {
  const sp = new URLSearchParams(parameters)
  sp.sort()
  return sp.toString().replace(/\+/g, '%20')
}

export const toDotNotation = (object: object, prefix: string) => {
  const result: { [key: string]: string | number | boolean } = {}
  function dotify(plainObject: object, currentKey?: string | number) {
    Object.entries(plainObject).forEach(([key, value]) => {
      const newKey = currentKey ? `${currentKey}.${key}` : key // joined key with dot
      if (value && typeof value === 'object') {
        dotify(value, newKey) // it's a nested object, so do it again
      } else {
        Object.assign(result, { [`${prefix}.${newKey}`]: value }) // it's not an object, so set the property
      }
    })
  }

  dotify(object)
  return result
}

export const cleanParameters = (parameters: Parameters): CleanParameters =>
  Object.entries(parameters)
    .filter(([, parameter]) => parameter !== undefined)
    .reduce((result, [key, parameter]) => {
      if (Array.isArray(parameter)) {
        parameter.forEach((parameterChild: string | number | object, index: number) => {
          if (typeof parameterChild === 'string' || !Number.isNaN(Number(parameterChild))) {
            Object.assign(result, { [`${key}.${index + 1}`]: String(parameterChild) })
          } else {
            Object.assign(result, toDotNotation(parameterChild as object, `${key}.${index + 1}`))
          }
        })
      } else {
        Object.assign(result, { [key]: String(parameter) })
      }

      return result
    }, {} as CleanParameters)

const defaultFetch = ({ url, method, headers, data }: Request): Promise<RequestResponse> =>
  axios({ method, url, headers, data })
    .then((response) => ({
      data: response.data,
      headers: response.headers,
    }))
    .catch((error) => {
      return Promise.reject(error.response.data)
    })

const parseResponse = <T>(response: RequestResponse): [T, RequestMeta] => {
  const responseData = parser.parse(response.data)

  return [
    responseData,
    {
      requestId: response.headers['x-mws-request-id'],
      timestamp: new Date(response.headers['x-mws-timestamp']),
      quotaMax: Number(response.headers['x-mws-quota-max']),
      quotaRemaining: Number(response.headers['x-mws-quota-remaining']),
      quotaResetOn: new Date(response.headers['x-mws-quota-resetson']),
    },
  ]
}

export class HttpClient {
  constructor(
    private options: MWSOptions,
    private fetch: <T>(meta: Request) => Promise<RequestResponse> = defaultFetch,
  ) {}

  public async request<TResource extends Resource, TRes>(
    method: HttpMethod,
    info: ResourceInfo<TResource>,
  ): Promise<[TRes, RequestMeta]> {
    const marketplaceUri = this.options.marketplace.webServiceUri

    const host = marketplaceUri.replace('https://', '')
    const url = `${marketplaceUri}/${info.resource}/${info.version}`

    const parameters = {
      AWSAccessKeyId: this.options.awsAccessKeyId,
      Action: info.action,
      MWSAuthToken: this.options.mwsAuthToken,
      SellerId: this.options.sellerId,
      SignatureMethod: 'HmacSHA256',
      SignatureVersion: '2',
      Timestamp: new Date().toISOString(),
      Version: info.version,
      ...cleanParameters(info.parameters),
    }

    const parametersForSigning = canonicalizeParameters(parameters)
    const queryStringToSign = `${method}\n${host}\n/${info.resource}/${info.version}\n${parametersForSigning}`

    const signature = sign(queryStringToSign, this.options.secretKey)
    const parametersWithSignature = { ...parameters, Signature: signature }

    const headers = {
      'user-agent': '@scaleleap/amazon-mws-api-sdk/1.0.0 (Language=JavaScript)',
    }

    const config =
      method === 'GET'
        ? {
            url: `${url}?${canonicalizeParameters(parametersWithSignature)}`,
            method,
            headers,
          }
        : {
            url,
            method,
            headers,
            data: canonicalizeParameters(parametersWithSignature),
          }

    try {
      const response = await this.fetch(config)

      // GetMatchingProductForId can return an Invalid UPC identifier error message to an otherwise successfully processed request (i.e. 200 status code)
      if (
        info.action === 'GetMatchingProductForId' &&
        response.data.includes('Invalid UPC identifier')
      ) {
        throw new InvalidUPCIdentifier(`${info.action} request failed`)
      }

      return parseResponse(response)
    } catch (error) {
      if (parser.validate(error) !== true) {
        throw error
      }

      const maybeResponse = MWSApiError.decode(parser.parse(error))

      if (maybeResponse.isRight()) {
        const response = maybeResponse.extract()
        const errorCode = response.ErrorResponse.Error.Code

        const errorMap = {
          InputStreamDisconnected,
          InvalidParameterValue,
          AccessDenied,
          InvalidAccessKeyId,
          SignatureDoesNotMatch,
          InvalidAddress,
          InternalError,
          // Subscriptions-specific
          'Internal Error': InternalError,
          QuotaExceeded,
          RequestThrottled,
          ResourceNotFound,
          ScheduledPackageAlreadyExists,
          RegionNotSupported,
          ScheduleWindowExpired,
          InvalidOrderState,
          PickupSlotNotAvailable,
          AccessToFeedProcessingResultDenied,
          ContentMD5Missing,
          ContentMD5DoesNotMatch,
          FeedCanceled,
          FeedProcessingResultNoLongerAvailable,
          FeedProcessingResultNotReady,
          InputDataError,
          InvalidFeedSubmissionId,
          InvalidFeedType,
          InvalidRequest,
          NonRetriableInternalError,
          RetriableInternalError,
          AccessToReportDenied,
          InvalidReportId,
          InvalidReportType,
          InvalidScheduleFrequency,
          ReportNoLongerAvailable,
          ReportNotReady,
          DependencyFatalException,
          DependencyRetriableException,
          DependencyUnauthorizedException,
          InternalErrorFatalException,
          InvalidInputFatalException,
        }

        const ErrorToThrow = errorMap[errorCode]

        throw enhanceError(new ErrorToThrow(`${info.action} request failed`), response)
      } else {
        throw error
      }
    }
  }
}
