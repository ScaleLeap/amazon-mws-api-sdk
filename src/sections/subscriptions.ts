import { boolean, Codec, enumeration, exactly, GetInterface, string } from 'purify-ts'

import { ParsingError } from '../error'
import { HttpClient, RequestMeta, Resource } from '../http'
import { ensureArray } from '../parsing'
import { getServiceStatusByResource } from './shared'

const SUBSCRIPTIONS_API_VERSION = '2013-07-01'

/**
 * Amazon docs list these as the only possible choices for each parameters
 */
export type DeliveryChannel = 'SQS'
export type AttributeKeyValueKeys = 'sqsQueueUrl'
export type NotificationType =
  | 'AnyOfferChanged'
  | 'FeedProcessingFinished'
  | 'FeePromotion'
  | 'FulfillmentOrderStatus'
  | 'ReportProcessingFinished'

enum NotificationTypeEnum {
  AnyOfferChanged = 'AnyOfferChanged',
  FeedProcessingFinished = 'FeedProcessingFinished',
  FeePromotion = 'FeePromotion',
  FulfillmentOrderStatus = 'FulfillmentOrderStatus',
  ReportProcessingFinished = 'ReportProcessingFinished',
}

const NotificationType = enumeration(NotificationTypeEnum)

interface MarketplaceIdAndDestinationOnlyParameters {
  MarketplaceId: string
  Destination: Destination
}

interface AttributeKeyValue {
  Key: AttributeKeyValueKeys
  Value: string
}
interface Destination {
  DeliveryChannel: DeliveryChannel
  AttributeList: AttributeKeyValue[]
}
interface SubscriptionActionParameters {
  MarketplaceId: string
  NotificationType: NotificationType
  Destination: Destination
}

type RegisterDestinationParameters = MarketplaceIdAndDestinationOnlyParameters

const RegisterDestinationResponse = Codec.interface({
  RegisterDestinationResponse: Codec.interface({
    RegisterDestinationResult: exactly(''),
  }),
})

type DeregisterDestinationParameters = MarketplaceIdAndDestinationOnlyParameters

const DeregisterDestinationResponse = Codec.interface({
  DeregisterDestinationResponse: Codec.interface({
    DeregisterDestinationResult: exactly(''),
  }),
})

interface ListRegisteredDestinationsParameters {
  MarketplaceId: string
  [key: string]: string
}
enum AttribueKeyValueKeysEnum {
  sqsQueueUrl = 'sqsQueueUrl',
}
const AttribueKeyValueKeys = enumeration(AttribueKeyValueKeysEnum)

const AttribueKeyValue = Codec.interface({
  Value: string,
  Key: AttribueKeyValueKeys,
})

enum DeliveryChannelEnum {
  SQS = 'SQS',
}

const DeliveryChannel = enumeration(DeliveryChannelEnum)

const Destination = Codec.interface({
  DeliveryChannel,
  AttributeList: ensureArray('member', AttribueKeyValue),
})

const ListRegisteredDestinations = Codec.interface({
  DestinationList: ensureArray('member', Destination),
})

type ListRegisteredDestinations = GetInterface<typeof ListRegisteredDestinations>

const ListRegisteredDestinationsResponse = Codec.interface({
  ListRegisteredDestinationsResponse: Codec.interface({
    ListRegisteredDestinationsResult: ListRegisteredDestinations,
  }),
})

type SendTestNotificationToDestinationParameters = MarketplaceIdAndDestinationOnlyParameters

const SendTestNotificationToDestinationResponse = Codec.interface({
  SendTestNotificationToDestinationResponse: Codec.interface({
    SendTestNotificationToDestinationResult: exactly(''),
  }),
})

interface Subscription {
  NotificationType: NotificationType
  Destination: Destination
  IsEnabled: boolean
}

interface CreateSubscriptionParameters {
  MarketplaceId: string
  Subscription: Subscription
}

const CreateSubscriptionResponse = Codec.interface({
  CreateSubscriptionResponse: Codec.interface({
    CreateSubscriptionResult: exactly(''),
  }),
})

type GetSubscriptionParameters = SubscriptionActionParameters

const Subscription = Codec.interface({
  NotificationType,
  Destination,
  IsEnabled: boolean,
})

const GetSubscription = Codec.interface({
  Subscription,
})

type GetSubscription = GetInterface<typeof GetSubscription>

const GetSubscriptionResponse = Codec.interface({
  GetSubscriptionResponse: Codec.interface({
    GetSubscriptionResult: GetSubscription,
  }),
})

type DeleteSubscriptionParameters = SubscriptionActionParameters

const DeleteSubscriptionResponse = Codec.interface({
  DeleteSubscriptionResponse: Codec.interface({
    DeleteSubscriptionResult: exactly(''),
  }),
})

interface UpdateSubscriptionParameters {
  MarketplaceId: string
  Subscription: Subscription
}

const UpdateSubscriptionResponse = Codec.interface({
  UpdateSubscriptionResponse: Codec.interface({
    UpdateSubscriptionResult: exactly(''),
  }),
})
export class Subscriptions {
  constructor(private httpClient: HttpClient) {}

  async updateSubscription(parameters: UpdateSubscriptionParameters): Promise<['', RequestMeta]> {
    const [response, meta] = await this.httpClient.request('POST', {
      resource: Resource.Subscriptions,
      version: SUBSCRIPTIONS_API_VERSION,
      action: 'UpdateSubscription',
      parameters: {
        MarketplaceId: parameters.MarketplaceId,
        Subscription: {
          NotificationType: parameters.Subscription.NotificationType,
          Destination: {
            DeliveryChannel: parameters.Subscription.Destination.DeliveryChannel,
            'AttributeList.member': parameters.Subscription.Destination.AttributeList,
          },
          IsEnabled: parameters.Subscription.IsEnabled,
        },
      },
    })

    return UpdateSubscriptionResponse.decode(response).caseOf({
      Right: (x) => [x.UpdateSubscriptionResponse.UpdateSubscriptionResult, meta],
      Left: (error) => {
        throw new ParsingError(error)
      },
    })
  }

  async deleteSubscription(parameters: DeleteSubscriptionParameters): Promise<['', RequestMeta]> {
    const [response, meta] = await this.httpClient.request('POST', {
      resource: Resource.Subscriptions,
      version: SUBSCRIPTIONS_API_VERSION,
      action: 'DeleteSubscription',
      parameters: {
        MarketplaceId: parameters.MarketplaceId,
        NotificationType: parameters.NotificationType,
        Destination: {
          DeliveryChannel: parameters.Destination.DeliveryChannel,
          'AttributeList.member': parameters.Destination.AttributeList,
        },
      },
    })

    return DeleteSubscriptionResponse.decode(response).caseOf({
      Right: (x) => [x.DeleteSubscriptionResponse.DeleteSubscriptionResult, meta],
      Left: (error) => {
        throw new ParsingError(error)
      },
    })
  }

  async getSubscription(
    parameters: GetSubscriptionParameters,
  ): Promise<[GetSubscription, RequestMeta]> {
    const [response, meta] = await this.httpClient.request('POST', {
      resource: Resource.Subscriptions,
      version: SUBSCRIPTIONS_API_VERSION,
      action: 'GetSubscription',
      parameters: {
        MarketplaceId: parameters.MarketplaceId,
        NotificationType: parameters.NotificationType,
        Destination: {
          DeliveryChannel: parameters.Destination.DeliveryChannel,
          'AttributeList.member': parameters.Destination.AttributeList,
        },
      },
    })

    return GetSubscriptionResponse.decode(response).caseOf({
      Right: (x) => [x.GetSubscriptionResponse.GetSubscriptionResult, meta],
      Left: (error) => {
        throw new ParsingError(error)
      },
    })
  }

  async createSubscription(parameters: CreateSubscriptionParameters): Promise<['', RequestMeta]> {
    const [response, meta] = await this.httpClient.request('POST', {
      resource: Resource.Subscriptions,
      version: SUBSCRIPTIONS_API_VERSION,
      action: 'CreateSubscription',
      parameters: {
        MarketplaceId: parameters.MarketplaceId,
        Subscription: {
          NotificationType: parameters.Subscription.NotificationType,
          Destination: {
            DeliveryChannel: parameters.Subscription.Destination.DeliveryChannel,
            'AttributeList.member': parameters.Subscription.Destination.AttributeList,
          },
          IsEnabled: parameters.Subscription.IsEnabled,
        },
      },
    })

    return CreateSubscriptionResponse.decode(response).caseOf({
      Right: (x) => [x.CreateSubscriptionResponse.CreateSubscriptionResult, meta],
      Left: (error) => {
        throw new ParsingError(error)
      },
    })
  }

  async sendTestNotificationToDestination(
    parameters: SendTestNotificationToDestinationParameters,
  ): Promise<['', RequestMeta]> {
    const [response, meta] = await this.httpClient.request('POST', {
      resource: Resource.Subscriptions,
      version: SUBSCRIPTIONS_API_VERSION,
      action: 'SendTestNotificationToDestination',
      parameters: {
        MarketplaceId: parameters.MarketplaceId,
        Destination: {
          DeliveryChannel: parameters.Destination.DeliveryChannel,
          'AttributeList.member': parameters.Destination.AttributeList,
        },
      },
    })

    return SendTestNotificationToDestinationResponse.decode(response).caseOf({
      Right: (x) => [
        x.SendTestNotificationToDestinationResponse.SendTestNotificationToDestinationResult,
        meta,
      ],
      Left: (error) => {
        throw new ParsingError(error)
      },
    })
  }

  async listRegisteredDestinations(
    parameters: ListRegisteredDestinationsParameters,
  ): Promise<[ListRegisteredDestinations, RequestMeta]> {
    const [response, meta] = await this.httpClient.request('POST', {
      resource: Resource.Subscriptions,
      version: SUBSCRIPTIONS_API_VERSION,
      action: 'ListRegisteredDestinations',
      parameters,
    })

    return ListRegisteredDestinationsResponse.decode(response).caseOf({
      Right: (x) => [x.ListRegisteredDestinationsResponse.ListRegisteredDestinationsResult, meta],
      Left: (error) => {
        throw new ParsingError(error)
      },
    })
  }

  async deregisterDestination(
    parameters: DeregisterDestinationParameters,
  ): Promise<['', RequestMeta]> {
    const [response, meta] = await this.httpClient.request('POST', {
      resource: Resource.Subscriptions,
      version: SUBSCRIPTIONS_API_VERSION,
      action: 'DeregisterDestination',
      parameters: {
        MarketplaceId: parameters.MarketplaceId,
        Destination: {
          DeliveryChannel: parameters.Destination.DeliveryChannel,
          'AttributeList.member': parameters.Destination.AttributeList,
        },
      },
    })

    return DeregisterDestinationResponse.decode(response).caseOf({
      Right: (x) => [x.DeregisterDestinationResponse.DeregisterDestinationResult, meta],
      Left: (error) => {
        throw new ParsingError(error)
      },
    })
  }

  async registerDestination(parameters: RegisterDestinationParameters): Promise<['', RequestMeta]> {
    const [response, meta] = await this.httpClient.request('POST', {
      resource: Resource.Subscriptions,
      version: SUBSCRIPTIONS_API_VERSION,
      action: 'RegisterDestination',
      parameters: {
        MarketplaceId: parameters.MarketplaceId,
        Destination: {
          DeliveryChannel: parameters.Destination.DeliveryChannel,
          'AttributeList.member': parameters.Destination.AttributeList,
        },
      },
    })

    return RegisterDestinationResponse.decode(response).caseOf({
      Right: (x) => [x.RegisterDestinationResponse.RegisterDestinationResult, meta],
      Left: (error) => {
        throw new ParsingError(error)
      },
    })
  }

  async getServiceStatus() {
    return getServiceStatusByResource(
      this.httpClient,
      Resource.Subscriptions,
      SUBSCRIPTIONS_API_VERSION,
    )
  }
}
