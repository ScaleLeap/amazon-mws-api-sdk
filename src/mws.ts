import { HttpClient } from './http'
import { FulfillmentInventory } from './sections/fulfillment-inventory'
import { Orders } from './sections/orders'
import { Products } from './sections/products/products'
import { Reports } from './sections/reports'
import { Sellers } from './sections/sellers'

export class MWS {
  private _sellers!: Sellers

  private _orders!: Orders

  private _fulfillmentInventory!: FulfillmentInventory

  private _products!: Products

  private _reports!: Reports

  constructor(private httpClient: HttpClient) {}

  get sellers() {
    if (!this._sellers) {
      this._sellers = new Sellers(this.httpClient)
    }

    return this._sellers
  }

  get orders() {
    if (!this._orders) {
      this._orders = new Orders(this.httpClient)
    }

    return this._orders
  }

  get fulfillmentInventory() {
    if (!this._fulfillmentInventory) {
      this._fulfillmentInventory = new FulfillmentInventory(this.httpClient)
    }

    return this._fulfillmentInventory
  }

  get products() {
    if (!this._products) {
      this._products = new Products(this.httpClient)
    }

    return this._products
  }

  get reports() {
    if (!this._reports) {
      this._reports = new Reports(this.httpClient)
    }

    return this._reports
  }
}
