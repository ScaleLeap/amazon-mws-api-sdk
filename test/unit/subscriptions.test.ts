import { amazonMarketplaces, HttpClient, ParsingError } from '../../src'
import { MWS } from '../../src/mws'
import { AttributeKeyValueKeys, DeliveryChannel } from '../../src/sections/subscriptions'
import { getFixture } from '../utils'

const httpConfig = {
  awsAccessKeyId: '',
  marketplace: amazonMarketplaces.CA,
  mwsAuthToken: '',
  secretKey: '',
  sellerId: '',
}

const headers = {
  'x-mws-request-id': '0',
  'x-mws-timestamp': '2020-05-06T09:22:23.582Z',
  'x-mws-quota-max': '1000',
  'x-mws-quota-remaining': '999',
  'x-mws-quota-resetson': '2020-04-06T10:22:23.582Z',
}

const createMockHttpClient = (fixture: string) =>
  new MWS(
    new HttpClient(httpConfig, () =>
      Promise.resolve({
        data: getFixture(fixture),
        headers,
      }),
    ),
  )

const mockMwsServiceStatus = createMockHttpClient('get_service_status')

const mockMwsFail = new MWS(
  new HttpClient(httpConfig, () => Promise.resolve({ data: '', headers: {} })),
)

const parsingError = 'Expected an object, but received a string with value ""'

describe('sellers', () => {
  const mockDestination = {
    DeliveryChannel: 'SQS' as DeliveryChannel,
    AttributeList: [
      {
        Key: 'sqsQueueUrl' as AttributeKeyValueKeys,
        Value: 'https%3A%2F%2Fsqs.us-east-1.amazonaws.com%2F51471EXAMPLE%2Fmws_notifications',
      },
    ],
  }

  describe('listRegisteredDestinations', () => {
    const parameters = {
      MarketplaceId: '',
    }

    it('returns a list of destinations if succesful', async () => {
      expect.assertions(1)

      const mockListRegisteredDestinations = createMockHttpClient(
        'subscriptions_list_registered_destinations',
      )

      expect(
        await mockListRegisteredDestinations.subscriptions.listRegisteredDestinations(parameters),
      ).toMatchSnapshot()
    })

    it("throws a parsing error when the response isn't valid", async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.subscriptions.listRegisteredDestinations(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('deregisterDestination', () => {
    const parameters = {
      MarketplaceId: '',
      Destination: mockDestination,
    }

    it('returns the standard response if deregistration is succesful', async () => {
      expect.assertions(1)

      const mockDeregisterDestination = createMockHttpClient('subscriptions_deregister_destination')

      expect(
        await mockDeregisterDestination.subscriptions.deregisterDestination(parameters),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the response is not valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.subscriptions.deregisterDestination(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('registerDestination', () => {
    const parameters = {
      MarketplaceId: '',
      Destination: mockDestination,
    }

    it('returns the standard response if registration is succesful', async () => {
      expect.assertions(1)

      const mockRegisterDestination = createMockHttpClient('subscriptions_register_destination')

      expect(
        await mockRegisterDestination.subscriptions.registerDestination(parameters),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the response is not valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.subscriptions.registerDestination(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('getServiceStatus', () => {
    it('returns a parsed model when the status response is valid', async () => {
      expect.assertions(1)

      expect(await mockMwsServiceStatus.subscriptions.getServiceStatus()).toMatchSnapshot()
    })

    it('throws a parsing error when the status response is not valid', async () => {
      expect.assertions(1)

      await expect(() => mockMwsFail.sellers.getServiceStatus()).rejects.toStrictEqual(
        new ParsingError(parsingError),
      )
    })
  })
})
