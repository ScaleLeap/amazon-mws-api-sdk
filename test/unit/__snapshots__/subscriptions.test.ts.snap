// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sellers createSubscription returns the standard response if creation is succesful 1`] = `
Array [
  "",
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`sellers deleteSubscription returns the standard response if delete is succesful 1`] = `
Array [
  "",
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`sellers deregisterDestination returns the standard response if deregistration is succesful 1`] = `
Array [
  "",
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`sellers getServiceStatus returns a parsed model when the status response is valid 1`] = `
Array [
  Object {
    "Status": "GREEN",
    "Timestamp": "2020-05-06T08:22:23.582Z",
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`sellers getSubscription returns a subscription if succesful 1`] = `
Array [
  Object {
    "Subscription": Object {
      "Destination": Object {
        "AttributeList": Array [
          Object {
            "Key": "sqsQueueUrl",
            "Value": "https://sqs.us-east-1.amazonaws.com/51471EXAMPLE/mws_notifications",
          },
        ],
        "DeliveryChannel": "SQS",
      },
      "IsEnabled": true,
      "NotificationType": "AnyOfferChanged",
    },
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`sellers listRegisteredDestinations returns a list of destinations if succesful 1`] = `
Array [
  Object {
    "DestinationList": Array [
      Object {
        "AttributeList": Array [
          Object {
            "Key": "sqsQueueUrl",
            "Value": "https://sqs.us-east-1.amazonaws.com/51471EXAMPLE/mws_notifications",
          },
        ],
        "DeliveryChannel": "SQS",
      },
    ],
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`sellers registerDestination returns the standard response if registration is succesful 1`] = `
Array [
  "",
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`sellers sendTestNotificationToDestination returns the standard response if testing is succesful 1`] = `
Array [
  "",
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;
